docs/design-documents/2020-02-Promtail-Push-API.md

```diff
@@ -0,0 +1,124 @@
+# Promtail Push API
+
+- Author: Robert Fratto (@rfratto)
+- Date: Feb 4 2020
+- Status: DRAFT
+
+Despite being an optional piece of software, Promtail provides half the power
+of Loki's story: log transformations, service discovery, metrics from logs,
+and context switching between your existing metrics and logs. Today, Promtail
+can only be operated to consume logs from very specific sources: files, journal,
+or syslog. If users wanted to write custom tooling to ship logs, the tooling
+has to bypass Promtail and push directly to Loki. This can lead users to
+reimplement functionality Promtail already provides, including its error retries
+and batching code.
+
+This document proposes a Push API for Promtail. By being compatible with the
+Loki Push API, the Promtail Push API can allow batches of logs to be processed
+at once for optimizing performance. Matching the Promtail API also allows users
+to transparently switch their push URLs from their existing tooling. Finally,
+a series of alternative solutions will be detailed.
+
+## Configuration
+
+Promtail will have a new target called HTTPTarget, configurable in the
+`scrape_config` array with the following schema:
+
+```yaml
+# Defines an HTTP target, which exposes an endpoint against the Promtail
+# HTTP server to accept log traffic.
+http:
+  # Defines the base URL for the push path, adding a prefix to the
+  # exposed endpoint. The final endpoint path is
+  # <base_url>loki/api/v1/push. If omitted, defaults to /.
+  #
+  # Multiple http targets with the same base_url must not exist.
+  base_url: /
+
+  # Map of labels to add to every log line passed through to the target.
+  labels: {}
+```
+
+### Considerations
+
+Users will be able to define multiple `http` scrape configs, but the base URL
```

Comment:
Would it make more sense to have pipelines available via suffix instead of base url?
`loki/api/v1/push/{A,B,C}` instead of `{A,B,C}/loki/api/v1/push`?.

Reply:
Hmm, could we still consider that Loki-compatible if it's a suffix?

Reply:
I'm not entirely sure if it's a problem.

One other thought is that we may want the libraries/applications that _push_ to promtail to be able to add their own labels dynamically (like in the case of adding http code labels to prometheus metrics in source code instrumentation). Consequently we'd need the ability to set labels both via the receiving endpoint on promtail's side as well as in the incoming payload. What do you think?

Reply:
I'm not sure this endpoint difference would actually be a problem, but I may be missing something here.

Reply:
I'm just thinking of the Grafana use-case where it adds the Loki API as a suffix, and allows a base URL with whatever prefix you want. If other endpoints had custom suffixes, Grafana wouldn't be able to function properly.

OTOH, Grafana doesn't push logs to Loki, so this isn't a real concern; I'm just assuming there might be other tooling that works this way.